src/GraphQL.MicrosoftDI/GraphQLBuilder.cs

```diff
@@ -0,0 +1,126 @@
+#nullable enable
+
+using System;
+using GraphQL.DI;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Options;
+using MSServiceLifetime = Microsoft.Extensions.DependencyInjection.ServiceLifetime;
+using ServiceLifetime = GraphQL.DI.ServiceLifetime;
+
+namespace GraphQL.MicrosoftDI
+{
+    /// <summary>
+    /// An implementation of <see cref="IGraphQLBuilder"/> which uses the Microsoft dependency injection framework
+    /// to register services and configure options.
+    /// </summary>
+    internal class GraphQLBuilder : GraphQLBuilderBase
+    {
+        private readonly IServiceCollection _services;
+        /// <summary>
+        /// Initializes a new instance for the specified service collection.
+        /// </summary>
+        /// <remarks>
+        /// Registers various default services via <see cref="GraphQLBuilderBase.Initialize"/>.
+        /// </remarks>
+        public GraphQLBuilder(IServiceCollection services)
+        {
+            _services = services ?? throw new ArgumentNullException(nameof(services));
+            services.AddOptions();
+            Initialize();
+        }
+
+        private static MSServiceLifetime TranslateLifetime(ServiceLifetime serviceLifetime)
+            => serviceLifetime switch
+            {
+                ServiceLifetime.Singleton => MSServiceLifetime.Singleton,
+                ServiceLifetime.Scoped => MSServiceLifetime.Scoped,
+                ServiceLifetime.Transient => MSServiceLifetime.Transient,
+                _ => throw new ArgumentOutOfRangeException(nameof(serviceLifetime))
+            };
+
+        /// <inheritdoc/>
+        public override IGraphQLBuilder Configure<TOptions>(Action<TOptions, IServiceProvider>? action = null)
+        {
+            this.TryRegister(services => services.GetService<IOptions<TOptions>>()?.Value ?? new TOptions(), DI.ServiceLifetime.Singleton);
```

Comment:
Interesting. So both `IOptions<T>` and 'raw' `T` are supported.

Reply:
Yes

Reply:
As long as the AddMemoryCache or similar methods call Configure for their options class, even with no delegate, then the mapping is installed so configuring the options outside of the builder will still work correctly even when MemoryCache doesn’t directly support IOptions. 

Also, if IOptions is not used, users can directly register T to pass through a configuration. Of course no configuration delegates get executed if they do so. 

It’s designed to be flexible and work with the user the way they want to work.